File: LibraryApp/Ee.Ebs.Application/Books/BookService.cs
using AutoMapper;
using Ee.Ebs.Application.Contracts.Books;
using Ee.Ebs.Application.Contracts.Books.DTOs;
using Ee.Ebs.Domain.Books;

namespace Ee.Ebs.Application.Books;

public class BookService : IBookService
{
    private readonly IBookRepository _bookRepository;
    private readonly IMapper _mapper;

    public BookService(IBookRepository bookRepository, IMapper mapper)
    {
        _bookRepository = bookRepository;
        _mapper = mapper;
    }
    
    public List<BookDto> GetAll()
    {
        var books = _bookRepository.GetAll();
        return _mapper.Map<List<BookDto>>(books);
    }
    
    public BookDto GetById(int id)
    {
        var book = _bookRepository.GetById(id);
        return _mapper.Map<BookDto>(book);
    }
    
    public void Insert(BookCreateDto dto)
    {
        var bookEntity = _mapper.Map<Book>(dto);
        bookEntity.IsActive = true; 
        _bookRepository.Insert(bookEntity);
    }
    
    public void Update(BookEditDto dto)
    {
        var bookEntity = _mapper.Map<Book>(dto);
        bookEntity.IsActive = true; 
        _bookRepository.Update(bookEntity);
    }
    
    public void Delete(int id)
    {
        var book = new Book { Id = id };
        _bookRepository.Delete(book);
    }

    public List<BookDto> GetAvailableBooks()
    {
        var availableBooks = _bookRepository.GetAvailableBooks();
        return _mapper.Map<List<BookDto>>(availableBooks);
    }
}

File: angular/src/app/features/books/services/book.service.ts
import { Injectable, Service } from '@angular/core';
import { BookCreateDto, BookDto, BookEditDto} from '../../books/models/book.dto'
import { Observable } from 'rxjs';
import { HttpClient } from '@angular/common/http';
import { environment } from '../../../../environments/environment.development';


@Injectable({
    providedIn: 'root'
})
export class BookService {
    private api_url = `${environment.apiUrl}/api/books`;

    constructor(private http: HttpClient){}

    getAll(): Observable<BookDto[]>{
        return this.http.get<BookDto[]>(this.api_url);
    }

    getById(id: number): Observable<BookDto>{
        return this.http.get<BookDto>(`${this.api_url}/${id}`);
    }

    getAvailableBooks(): Observable<BookDto[]>{
        return this.http.get<BookDto[]>(`${this.api_url}/available`);
    }

    getBooksByCategory(categoryId: number): Observable<BookDto[]> {
        return this.http.get<BookDto[]>(`${this.api_url}/book-category/${categoryId}`);
  }

    Insert(book: BookCreateDto): Observable<BookDto>{
        return this.http.post<BookDto>(this.api_url, book);
    }

    Update(id: number, book: BookEditDto): Observable<BookDto>{
        return this.http.put<BookDto>(`${this.api_url}/${id}`, book);
    }

    Delete(id: number): Observable<void>{
        return this.http.delete<void>(`${this.api_url}/${id}`);
    }
}
